src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -216,24 +222,81 @@ protected static class InternalKey extends NativeResource {
     T getPublicKey();
   }
 
-  // The Key interface requires us to be serializable.
-  // However, we are not easily serializable because we have native state which won't be captured.
-  // So, we prevent any attempt to serialize ourselves by throwing an exception.
-  // If we discover that we actually need to support this functionality, we can manage it through
-  // serializing an
-  // encoded copy of the key and decoding that upon deserialization.
-  // Doing that will require changing out InternalKey from being a final field to a transient one
-  // and this feels
-  // like a bad tradeoff if we can avoid it.
+  /*
+   * Java Keys are Serializable but we cannot use the trivial serialization of an EvpKey.
+   * - They contain a pointer to native memory (and associated native memory).
+   *   This obviously cannot remain valid after being serialized/deserialized.
+   * - The pointer to native memory is final to ensure there is no risk of it being invalid.
+   *   This means that we cannot just fix it up when deserializing.
+   *
+   * The trivial solution would be to make the pointer non-final, but that introduces a greater risk of serious bugs.
+   *
+   * Instead, we use writeReplace() and readResolve() to store our information in a dedicated format for serialization.
+   * - writeReplace() on the object to be serialized (an EvpKey) returns the *real* object to be serialized.
+   *   In our case, we return an instance of SerializedKey which contains the minimal information to properly save and retrieve a key.
+   * - readResolve() is called on the object which is *actually* serialized (and thus being deserialized) and returns the object that users care about.
+   *   In our case, SeralizedKey.readResolve() returns an appropriate instance of EvpKey.
+   *
+   * So, the entire (hidden from the user) flow goes like this.
+   * 1.  User tries to serialize an EvpKey
+   * 2.  Java detects that EvpKey.writeReplace() exists and calls it
+   * 3.  EvpKey.writeReplace() returns an instance of SerializedKey
+   * 4.  Java actually serializes SerializedKey
+   * 5.  (Some time passes)
+   * 6.  User tries to deserialize the bytes
+   * 7.  Java detects that the bytes contain a serialized SerializedKey and so deserializes it.
+   * 8.  Java detects that SerializedKey.readResolve() exists and calls it
+   * 9.  SerializedKey.readResolve() creates and returns an instance of EvpKey
+   * 10. User gets an instance of EvpKey and never realizes that the above complexity exists
+   */
+  Object writeReplace() throws ObjectStreamException {
```

Comment:
I think that we cannot make it private. From the [Serializable documentation](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/io/Serializable.html) (emphasis mine):

> This writeReplace method is invoked by serialization if the method exists and it would be **accessible from a method defined within the class of the object being serialized**.

Thus, if we make it `private`, then it isn't accessible when we need to serialize any of the concrete implementations of `EvpKey`.